tempesta_fw/http_sticky.c

```diff
@@ -78,67 +78,87 @@ tfw_http_sticky_send_302(TfwHttpMsg *hm)
 	return 0;
 }
 
-/*
- * Find a specific non-special header field in an HTTP message.
- *
- * This function assumes that the header field name is stored
- * in TfwStr{} after an HTTP message is parsed.
- */
-static TfwStr *
-tfw_http_field_raw(TfwHttpMsg *hm, const char *field_name, size_t len)
+static bool
+search_cookie(TfwPool *pool, const TfwStr *cookie, TfwStr *val)
 {
-	int i;
-
-	for (i = TFW_HTTP_HDR_RAW; i < hm->h_tbl->size; i++) {
-		TfwStr *hdr_field = &hm->h_tbl->tbl[i];
-		if (tfw_str_eq_cstr(hdr_field, field_name, len,
-				    TFW_STR_EQ_PREFIX | TFW_STR_EQ_CASEI))
-			return hdr_field;
-	}
-
-	return NULL;
-}
-
-static int
-tfw_http_field_value(TfwHttpMsg *hm, const TfwStr *field_name, TfwStr *value)
-{
-	char *buf, *ptr;
-	size_t len;
-	TfwStr *hdr_field;
-
-	hdr_field = tfw_http_field_raw(hm, field_name->ptr, field_name->len);
-	if (hdr_field == NULL) {
-		return 0;
-	}
-
-	if (TFW_STR_PLAIN(hdr_field)) {
-		buf = (char *)hdr_field->ptr + field_name->len;
-		len = hdr_field->len - field_name->len;
-
-		while (len > 0 && isspace(*buf)) {
-			buf++;
-			len--;
-		}
-		while (len > 0 && isspace(*(buf + len - 1)))
-			len--;
+	const char *const cstr = tfw_cfg_sticky.name.ptr;
+	const unsigned int clen = tfw_cfg_sticky.name.len + 1;
+
+	const TfwStr *chunk, *next;
+	TfwStr *s = NULL;
+	unsigned int n = TFW_STR_CHUNKN(cookie) + 1;
+	TfwStr tmp;
+	enum {
+		StateName,
+		StateSearchVal,
+		StateVal,
+		StateValChunks,
+	} state = StateName;
+
+	BUG_ON(!TFW_STR_PLAIN(&tfw_cfg_sticky.name));
+
+	TFW_STR_INIT(&tmp);
+	tmp.flags = __TFW_STR_COMPOUND;
+
+	TFW_STR_FOR_EACH_CHUNK(chunk, cookie, {
```

Comment:
Cookies can be large and this piece of code could be really hot (this is the intention of #199).

Never use the dummy FSM style in performance critical code, please read http://natsys-lab.blogspot.ru/2014/11/the-fast-finite-state-machine-for-http.html for more details.

Moreover, TfwStr containing Cookie is never plain so you don't need generic TfwStr loop provided by TFW_STR_FOR_EACH_CHUNK() and can use very simple common for() loop.